README.md
# Crowd Rest

A ruby client for [Atlassian's Crowd REST API](http://confluence.atlassian.com/display/CROWDDEV/Crowd+REST+APIs).

## Installation

    gem install crowd_rest

## Usage

    require 'crowd_rest'
    
    CrowdRest.config do |c|
      c.crowd_url = "http://127.0.0.1:8095"
      c.app_name = "demo"
      c.app_pass = "demo_pass"
    end
    
    # unsuccessful login
    response = CrowdRest::Session.create("baduser", "badpass")
    response.code # => 400
    response.reason # => INVALID_USER_AUTHENTICATION
    
    # successful log in
    response = CrowdRest::Session.create("gooduser", "goodpass")
    response.code # => 201
    token = response.token # => the crowd sso token

    # check for existing login session
    response = CrowdRest::Session.find(token)
    response.code # => 200
    
    # get the user associated with a login session
    response = CrowdRest::Session.find(token, :include => :user)
    response.user.name # => "gooduser"

*We use [semantic versioning](http://semver.org/). You should too.*


CHANGELOG.md
# 2016-01-05: v0.1.1

* Bump HTTParty Gem Version requirement to 0.13

# 2013-03-16: v0.1.0

* Add support for user session destruction and token validation and
  invalidation (@kunklejr).

# 2011-05-18: v0.0.2

* The returned user from CrowdRest::Session.find(token, :include => :user) is
  now a CrowdRest::User object instead of a vanilla OpenStruct.

# 2011-05-17: v0.0.1

* Initial version. Supports login with CrowdRest::Session.create and finding
  current user/session with CrowdRest::Session.find


spec/spec_helper.rb
require "rubygems"
require "bundler/setup"

require "rspec"
require "vcr"

require "crowd_rest"

CrowdRest.config do |c|
  c.crowd_url = "http://127.0.0.1:8095"
  c.app_name = "demo"
  c.app_pass = "demo_pass"
end

VCR.config do |c|
  c.cassette_library_dir = File.expand_path("cassette_library", File.dirname(__FILE__))
  c.stub_with :fakeweb
  c.ignore_localhost = false
end

if ENV['DONT_FAKE']
  puts "VCR: Will record all http requests made during this test run..."
  record_type = :all
  at_exit { VCR.eject_cassette }
else
  puts "VCR: Using previously recorded http requests. " +
       "Run with DONT_FAKE=1 to hit local crowd server and record new requests."
  record_type = :none
end

VCR.insert_cassette('test_data', :record => record_type)

RSpec.configure do |config|
  config.extend VCR::RSpec::Macros
end


spec/crowd_rest_spec.rb
require "spec_helper"

describe CrowdRest do
  context "given valid app credentials" do
    it "will not get an '401 Unauthorized' response" do
      # valid params are set in spec_helper
      response = CrowdRest.get("/search")
      response.code.should_not == 401
    end
  end

  context "given invalid app credentials" do
    before(:all) do
      @old_app_name = CrowdRest.app_name
      @old_app_pass = CrowdRest.app_pass
      CrowdRest.app_name = "badname"
      CrowdRest.app_pass = "badpass"
    end
    
    after(:all) do
      CrowdRest.app_name = @old_app_name
      CrowdRest.app_pass = @old_app_pass
    end
    
    it "will get an '401 Unauthorized' response" do
      response = CrowdRest.get("/search")
      response.code.should == 401      
    end
  end
end


spec/crowd_rest/user_spec.rb
require 'spec_helper'

describe CrowdRest::User do
  before do
    login = CrowdRest::Session.create("crowduser", "crowdpass")
    @user = CrowdRest::Session.find(login.token, :include => :user).user
  end
  
  it "has the expected attributes for a user" do
    @user.first_name.should == "Crowd"
    @user.last_name.should == "Test"
    @user.display_name.should == "Crowd Test"
    @user.email.should == "crowduser@test.com"
    @user.name.should == "crowduser"
  end
end


lib/crowd_rest/session.rb
require 'ostruct'

module CrowdRest
  class Session
    def self.create(username, password)
      body = "<authentication-context>
        <username>#{username}</username>
        <password>#{password}</password>
      </authentication-context>"
      response = CrowdRest.post("/session", :body => body)
      normalize_response(response, 201) do |successful_response|
        successful_response.token = response['session']['token']
      end
    end

    def self.find(token, options = {})
      request_user = options[:include] && options[:include] == :user
      path = "/session/#{token}"
      path << "?expand=user" if request_user
      response = CrowdRest.get(path)
      normalize_response(response) do |successful_response|
        user = response['session']['user']
        successful_response.user = CrowdRest::User.new(user)
      end
    end

    def self.destroy(username, options = {})
      path = "/session?username=#{username}"
      path << "&except=#{options.except}" if options[:except]
      response = CrowdRest.delete(path)
      normalize_response(response, 204)
    end

    def self.validate(token, validation_factors = {})
      path = "/session/#{token}"
      body = "<validation-factors>"
      validation_factors.each do |name, value|
        body << "<validation-factor>
          <name>#{name}</name>
          <value>#{value}</value>
        </validation-factor>"
      end
      body << "</validation-factors>"
      response = CrowdRest.post(path, :body => body)
      normalize_response(response, 200) do |successful_response|
        successful_response.token = response['session']['token']
      end
    end

    def self.invalidate(token)
      path = "/session/#{token}"
      response = CrowdRest.delete(path)
      normalize_response(response, 204)
    end

    private
    def self.normalize_response(response, success_code = 200)
      attributes = {
        :code => response.code,
        :body => response.body,
        :reason => response['error'] ? response['error']['reason'] : response['reason'] || nil,
        :message => response['error'] ? response['error']['message'] : response['message'] || nil
      }

      norm_response = OpenStruct.new(attributes)
      yield(norm_response) if block_given? && response.code == success_code
      norm_response
    end
  end
end
